django/db/models/sql/compiler.py

```diff
@@ -320,18 +320,20 @@ def get_order_by(self):
                     OrderBy(Ref(col, self.query.annotation_select[col]), descending=descending),
                     True))
                 continue
-            if col in self.query.annotations:
+
+            if self.query.combinator and self.select:
+                # Don't use the resolved annotation because other
+                # combinated queries might define it differently.
+                expr = F(col)
+                order_by.append((OrderBy(expr, descending=descending), False))
+                continue
+            elif col in self.query.annotations:
```

Comment:
This will skip logic for `extra` columns. I added a test coverage and moved the fix to a different branch.